sim/midas/targetutils/src/main/scala/midas/xdc/RAMStyleHint.scala

```diff
@@ -0,0 +1,73 @@
+// See LICENSE for license details.
+
+package midas.targetutils.xdc
+
+import chisel3.experimental.{ChiselAnnotation}
+import firrtl.annotations.ReferenceTarget
+import midas.targetutils.xdc.{XDCFiles, XDCAnnotation}
+
+sealed trait RAMStyle
+
+/**
+  * Some rough guidance, based on Ultrascale+, is provided in the scala doc for
+  * each hint. Consult the Xilinx UGs for your target architecture and the
+  * synthesis UG (UG901).
+  */
+object RAMStyles {
+  /**
+    *  From UG901 (v2020.2): Instructs the tool to use the UltraScale+ URAM primitives.
+    *
+    *  URAMs are 288Kb, 72b wide, 4096 deep.
+    */
+  case object ULTRA extends RAMStyle
+
+  /**
+    *  From UG901 (v2020.2): Instructs the tool to infer RAMB type components
+    *
+    *  In Ultrascale+ (and older families), BRAMs are 36Kb, and have flexible aspect
+    *  ratio: 1b x 64K to 72b x 512
+    */
+  case object BRAM extends RAMStyle
+
+  /** From UG901 (v2020.2): Instructs the tool to infer registers instead of RAMs. */
+  case object REGISTERS extends RAMStyle
+
+  /** From UG901 (v2020.2): Instructs the tool to infer the LUT RAMs. */
+  case object DISTRIBUTED extends RAMStyle
+
+  /** Introduced in v2020.2. From UG901 (v2020.2):
+    *
+    * Instructs the tool to infer a combination of RAM types designed to minimize the
+    * amount of space that is unused.
+    *
+    * Note: This should probably be avoided for non-emulation-class
+    * FPGAs (e.g., VU19P), which tend to have rich embedded memory resources, as
+    * these designs tend to be under heavy LUT pressure. Here sparsely using
+    * BRAM / URAM resources insead of a space-optimal hybrid is desirable.
+    */
+  case object MIXED extends RAMStyle
+}
```

Comment:
What happens if/when we support platforms other than Xilinx?  Do you have thoughts on how to extend this?  Doesn't have to be done in this PR.

Reply:
I did think about it. The big question is whether we want the specialization to happen at elaboration time or RTL compile time. 

If we choose the prior, i think it would be natural to add more classes of style which would imply a different constraint string and different output file. There could be companion objects for each arch type for doing arch-specific annotations, and a general purpose one which could use an arch-agnostic hint, or a heuristic about the size / parameterization. 

If we choose the latter, well the whole annotation interface would change, but the pass could do global optimizations in RAM selection. This strikes me as somewhat fraught though, since Vivado does the same thing, and its non trivial to reason through the interactions with other Vivado compile directives. 

Really, my intuition tells me that we should only really use these to coax vivado into doing something non-default (i.e., deploy URAMs even if they are going to be very-sparsely filled), but otherwise remain hands off. If that's the case -- having extremely precise arch-specific annotations is probably preferable.